clang/lib/Sema/SemaCast.cpp

```diff
@@ -2532,8 +2532,7 @@ static TryCastResult TryReinterpretCast(Sema &Self, ExprResult &SrcExpr,
                        diag::err_bad_reinterpret_cast_small_int;
       return TC_Failed;
     }
-    Kind = SrcIsCap && !DestType->isIntCapType() ? CK_CHERICapabilityToAddress
-                                                 : CK_PointerToIntegral;
+    Kind = CK_PointerToIntegral;
```

Comment:
This is rather academic given `SrcType->isNullPtrType()` above, but seems more sensible to use, though `CastOperation::checkCapabilityToIntCast` returns `CK_NoOp` for this so perhaps that's what we want too?

Reply:
I'm not sure `CK_NoOp` will work correctly since it might confuse later checks (while it's a no-op in terms of codegen, it isn't semantically). I think using `CK_PointerToIntegral` unconditionally makes more sense.

Reply:
Do we have a test that casts `nullptr` -> integer?

Reply:
Well if it works for `CastOperation::checkCapabilityToIntCast` whenever `SrcType->isNullPtrType()`, surely it would work here too?

Reply:
But no, we don't currently test coverage for that, clang/test/CodeGenCXX/cheri/cheri-nullptr.cpp is pretty limited.

Reply:
Ok, it's CK_PointerToIntegral upstream, so yeah, CK_NoOp is wrong. And `CastOperation::checkCapabilityToIntCast` is really just a checking thing, it's not where the real casts are created for C.